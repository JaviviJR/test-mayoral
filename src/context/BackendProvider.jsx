import { useEffect, useState } from "react";
import { BackendContext } from "./BackendContext";

const URI_PRODUCTS = 'http://localhost:9005/products';

function BackendProvider({ children }) {
    const [products, setProducts] = useState([]);

    useEffect(() => {
        getProducts();
    }, []);

    const getProducts = async (queryString) => {
        const uri = queryString ? `${URI_PRODUCTS}?${queryString}` : URI_PRODUCTS;
        const response = await fetch(uri);
        const data = await response.json();
        setProducts(data);
    }

    const orderProducts = (order) => {
        const productSorted = [...products];

        if (order === 'ASC') {
            productSorted.sort((a, b) => (a.offer ? a.offer : a.price) > (b.offer ? b.offer : b.price));
        } else if (order === 'DESC') {
            productSorted.sort((a, b) => (a.offer ? a.offer : a.price) < (b.offer ? b.offer : b.price));
        }

        setProducts(productSorted);
        
    }

    const searchProduct = (search) => {
        const queryString = `name_like=${search}`;
        getProducts(queryString);
    }

    return (
        <BackendContext.Provider value={{
            products,
            orderProducts,
            searchProduct
        }}>
            { children }
        </BackendContext.Provider>
    );
}

export default BackendProvider;