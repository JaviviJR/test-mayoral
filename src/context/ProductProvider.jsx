import { useState } from "react";
import { ProductContext } from "./ProductContext";

const URI_PRODUCTS = 'http://localhost:9005/products';

function ProductProvider({ children }) {
    const [products, setProducts] = useState([]);

    const getProducts = async (search) => {
        const uri = search ? `${URI_PRODUCTS}?name_like=${search}` : URI_PRODUCTS;
        const response = await fetch(uri);
        const data = await response.json();
        setProducts(data);
    }

    const orderProducts = (order) => {
        const productSorted = [...products];

        if (order === 'ASC') {
            productSorted.sort((a, b) => (a.offer ? a.offer : a.price) > (b.offer ? b.offer : b.price));
        } else if (order === 'DESC') {
            productSorted.sort((a, b) => (a.offer ? a.offer : a.price) < (b.offer ? b.offer : b.price));
        }

        setProducts(productSorted);
        
    }

    return (
        <ProductContext.Provider value={{
            products,
            orderProducts,
            getProducts
        }}>
            { children }
        </ProductContext.Provider>
    );
}

export default ProductProvider;