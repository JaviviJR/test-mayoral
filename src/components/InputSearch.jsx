import { useContext, useEffect, useState } from "react";
import { ProductContext } from "../context/ProductContext";

function InputSearch() {
    const { getProducts } = useContext(ProductContext);
    const [search, setSearch] = useState('');

    useEffect(() => {
        const timeout = setTimeout(() => {
            getProducts(search);
        }, 500);
        return () => clearTimeout(timeout);
    }, [search]);

    return (
        <input
            type="text"
            className="form-control"
            placeholder="Buscar"
            name="search"
            value={ search }
            onChange={ (e) => setSearch(e.target.value) }
        />
    );
}

export default InputSearch;