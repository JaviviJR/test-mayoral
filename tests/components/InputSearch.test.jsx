import { render, screen, fireEvent, waitFor } from "@testing-library/react"
import InputSearch from "../../src/components/InputSearch";
import { ProductContext } from "../../src/context/ProductContext";

describe('test <InputSearch />', () => {
    const getProducts = jest.fn();
    
    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('debe coincidir con el snapshot', () => {
        const { container } = render(
            <ProductContext.Provider value={{
                getProducts: getProducts
            }}>
                <InputSearch />
            </ProductContext.Provider>
        );
        expect(container).toMatchSnapshot();
    })

    test('debe llamar getProducts() al escribir en el input', async () => {
        render(
            <ProductContext.Provider value={{
                getProducts
            }}>
                <InputSearch />
            </ProductContext.Provider>
        );

        const input = screen.getByRole('textbox');
        fireEvent.change(input, { target: { value: 'test' } });
        
        await waitFor(() => {
            expect(getProducts).toHaveBeenCalledTimes(1);
        });
    })
})
