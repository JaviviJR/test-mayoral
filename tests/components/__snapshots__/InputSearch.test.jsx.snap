// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <InputSearch /> debe coincidir con el snapshot 1`] = `
<div>
  <input
    class="form-control"
    name="search"
    placeholder="Buscar"
    type="text"
    value=""
  />
</div>
`;
